File: keys and rooms.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        ArrayList<ArrayList<Integer>>g=new ArrayList<ArrayList<Integer>>();
        int n=rooms.size();
        for(int i=0;i<=n;i++){
            g.add(new ArrayList<Integer>());
        }
        for(int i=0;i<rooms.size();i++){
            for(int j=0;j<rooms.get(i).size();j++){
                int u=i,v=rooms.get(i).get(j);
                g.get(u).add(v);
            }
        }
    boolean visit[]=new boolean[n+1];
     visit[0]=true;
     Queue<Integer>q=new LinkedList<Integer>();
     q.add(0);
     while(q.size()>0){
         int sz=q.size();
         while(sz-->0){
             int cur=q.peek();q.remove(q.peek());
             for(int v:g.get(cur)){
                 if(visit[v]==false){
                     q.add(v);
                     visit[v]=true;
                 }
             }
         }
     }
     for(int i=0;i<n;i++){
         if(visit[i]==false){return false;}
     }
     return true;

    
    }
}

File: keys and rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        int n=rooms.size();
        queue<int>q;
        vector<int>g[n+1];
        for(int i=0;i<rooms.size();i++){
            for(int j=0;j<rooms[i].size();j++){
                int u=i,v=rooms[i][j];
                g[u].emplace_back(v);
                
            }
        }
        bool visit[n+1];memset(visit,false,sizeof(visit));
        visit[0]=true;
        q.push(0);
        unordered_map<int,int>mp;
        while(q.size()){
            int sz=q.size();
            while(sz-->0){
                int cur=q.front();q.pop();
                 for(int v:g[cur]){
                     if(!visit[v]){
                         q.push(v);
                         visit[v]=true;
                     }
                 }
            }
        }
        for(int i=0;i<n;i++){if(!visit[i]){return false;}}
        return true;
    }
};